python/ray/_private/usage/usage_lib.py

```diff
@@ -124,3 +154,81 @@ def get_cluster_metadata(gcs_client, num_retries) -> dict:
             num_retries=num_retries,
         )
     )
+
+
+def generate_report_data(cluster_metadata: dict) -> dict:
+    """Generate the report data.
+
+    Note that we don't perform schema validation.
+
+    The server side is in charge of schema validation, and the schema
+    will be available via public Pypi wheel. The e2e validation will
+    be done from the test.
+
+    cluster_metadata (dict): The cluster metadata of the system generated by
+        `_generate_cluster_metadata`.
+    """
+    data = cluster_metadata.copy()
+    data["collect_timestamp_ms"] = int(time.time() * 1000)
+    return data
+
+
+def _write_usage_data(data: dict, dir_path: str) -> None:
+    """Write the usage data to the directory.
+
+    Params:
+        data (dict): Data to report
+        dir_path (Path): The path to the directory to write usage data.
+    """
+    assert _usage_stats_enabled()
+    # Atomically update the file.
+    dir_path = Path(dir_path)
+    destination = dir_path / usage_constant.USAGE_STATS_FILE
+    temp = dir_path / f"{usage_constant.USAGE_STATS_FILE}.tmp"
+    with temp.open(mode="w") as json_file:
+        json_file.write(json.dumps(data))
+    temp.rename(destination)
+
+
+def _report_usage_data(url: str, data: dict) -> None:
+    """Report the usage data to the usage server.
+
+    Params:
+        url (str): The URL to update resource usage.
+        data (dict): Data to report.
+
+    Raises:
+        requests.HTTPError if requests fails.
+    """
+    assert _usage_stats_enabled()
+    r = requests.request(
+        "POST",
+        url,
+        headers={"Content-Type": "application/json"},
+        json=data,
+        timeout=10,
+    )
+    r.raise_for_status()
+    return r
+
+
+async def write_usage_data_async(data: dict, dir_path: str, executor: Executor) -> None:
+    loop = asyncio.get_running_loop()
+    try:
+        await loop.run_in_executor(executor, _write_usage_data, data, dir_path)
+    except Exception as e:
+        logger.info(
+            f"Failed to write usage stats to {dir_path}. "
+            f"This is harmless and won't affect Ray. Error message: {e}"
+        )
+
+
+async def report_usage_data_async(url: str, data: dict, executor: Executor) -> None:
```

Comment:
Hmm I actually tried, and one of my concerns is how to cleaning up the thread when the client shuts down. I feel like just dynamically create/delete it like the current implementation is simpler because we don't need to explicitly manage the threads. 

I moved the functions to a UsageStatsClient class without self.executor. Lmk if you think it is better.

Reply:
You don't need to shut down the thread, just mark it as a daemon and let it life forever in a loop.